### PR TITLE
add component to display contributors

### PR DESCRIPTION
**Pull request by grmbyrn**
* in `routes/(home)/+page.server.ts`, fetch contributors data from GitHub API to use in Torrust repos.
  * Create a `Contributor` type for data coming from GitHub API and store it in `'$lib/utils/types'` to be imported
  * Create an array of repo names and store it in `'$lib/constants/constants'` to be imported and reactively mapped over together with a baseURL
  * Create a Personal Access Token, store it in `.env` and store it in `	const token = import.meta.env.VITE_GITHUB_TOKEN;` to order to increase GitHub API rate limits.
  * Fetch contributor data from multiple GitHub repository URLs concurrently using Promise.all. It sends authenticated GET requests (with a token in the headers) to each URL and parses the JSON response for all the requests.
  * Flatten the array of contributor arrays into a single array of all contributors.
  * Remove duplicate contributors from the all contributors array.
  * Pass array into `routes/(home)/+page.svelte`, which passes it into component `Contributors.svelte` in order to loop over and display each contributor on the home page.

* On each of the `(pages)` routes there wasn't responsiveness to make Table of Contents appear above the text on smaller screens and to the left on wider screens, so I added that to each.

### CODE REVIEW

File: src/lib/constants/constants.ts [hunk 1]
```diff
@@ -706,3 +706,30 @@ export const aboutTableData = [
...
+	'torrust-hash2torrent',
+	'torrust-index-gui-user-guide',
+	'containerizing-rust-apps-examples',
+	'torrust-serde-bencode-archive',
+	'torrust-index-archive',
+	'torrust-parse-torrent',
+	'torrents-importer-sample',
+	'torrust-documentation',
+	'torrust-docker'
+];
```
Comment (josecelano) [inline src/lib/constants/constants.ts]: Hi @grmbyrn I guess you can also get the list of repos from the API.

File: src/routes/(home)/+page.server.ts [hunk 2]
```diff
@@ -8,8 +10,28 @@ export async function load() {
 		.sort((a, b) => new Date(b.meta.date).getTime() - new Date(a.meta.date).getTime())
 		.slice(0, 6);
 
+	const baseURL = 'https://api.github.com/repos/torrust/';
+	const token = import.meta.env.VITE_GITHUB_TOKEN;
```
Comment (josecelano) [inline src/routes/(home)/+page.server.ts]: Hi, @grmbyrn, since the app is hosted statically on GitHub Pages, the environment variable's value (`VITE_GITHUB_TOKEN`) would be embedded directly into the build output (JavaScript code).

This means anyone can view it by inspecting the network requests or source code of your hosted app.

On the other hand, I think we don't need to do it every time a user loads the homepage. For now, we can build the data at build time because we are making many changes. We could create a file like `src/lib/constants/constants.ts` or maybe add a new static JSON file with the info. We need to find the easiest way to do it.

In the future, it the page is not updated frequently we can add a workflow to update that list (rebuild and publish) preiodically.

**Review by josecelano — CHANGES_REQUESTED** [review]

File: src/routes/(home)/+page.server.ts (hunk 2 above)
Comment (josecelano) [reply src/routes/(home)/+page.server.ts]: Hi @grmbyrn, as we discussed yesterday in the meeting, we can remove the token or use it only when it's provided (development environment). I did not realize the API rate limit is per IP, so it would be acceptable to do it at runtime. As @da2ce7 suggested, we can use a cache to store the result per user. I guess the cache could be implemented with https://svelte.dev/docs/svelte/stores.

Comment (grmbyrn) [reply src/routes/(home)/+page.server.ts]: > Hi @grmbyrn, as we discussed yesterday in the meeting, we can remove the token or use it only when it's provided (development environment). I did not realize the API rate limit is per IP, so it would be acceptable to do it at runtime. As @da2ce7 suggested, we can use a cache to store the result per user. I guess the cache could be implemented with https://svelte.dev/docs/svelte/stores.

Hi @josecelano I was able to cache the data coming from the GitHub API with a store. The initial request is within the 60 requests an hour permitted by the API. After that it fetches the requests from the store. At the moment the cache expires every hour.

**Review by josecelano — CHANGES_REQUESTED** [review]
Hi @grmbyrn, when you run it locally you get this error:

image

```output
  ➜  Local:   http://localhost:5173/
  ➜  Network: http://192.168.1.37:5173/
  ➜  Network: http://192.168.1.33:5173/
  ➜  press h + enter to show help
Fetching data from API
Rate limit: 60, Remaining: 0 10:46 AM
Rate limit resets at: Mon Dec 02 2024 10:53:01 GMT+0000 (Western European Standard Time)
Failed to fetch from https://api.github.com/orgs/torrust/repos: rate limit exceeded
Error: Failed to fetch from https://api.github.com/orgs/torrust/repos
    at Module.fetchWithCache (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/src/lib/utils/cache.ts:40:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async load (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/src/routes/(home)/+page.server.ts:13:16)
    at async Module.load_server_data (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:61:17)
    at async eval (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/page/index.js:140:13)
Fetching data from API
Rate limit: 60, Remaining: 0 10:46 AM
Rate limit resets at: Mon Dec 02 2024 10:53:01 GMT+0000 (Western European Standard Time)
Failed to fetch from https://api.github.com/orgs/torrust/repos: rate limit exceeded
Error: Failed to fetch from https://api.github.com/orgs/torrust/repos
    at Module.fetchWithCache (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/src/lib/utils/cache.ts:40:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async load (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/src/routes/(home)/+page.server.ts:13:16)
    at async Module.load_server_data (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:61:17)
    at async Promise.all (index 2)
    at async Module.render_data (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/data/index.js:99:17)
    at async resolve (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/respond.js:436:17)
    at async Module.respond (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/respond.js:322:20)
    at async file:///home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:524:22
```
I guess it's because I restarted the server and the data is not in the store anymore. I think we should not show the contributors list in that case.

I would also add the list to the community page.

Finally, I would remove console logs for production, especially when it's only for debugging.

### TODO

- [ ] Fix problem when get too many requests error but the store is empty. I think we could have a default value in this case with the list of contributor. We can update that default value from time to time.
- [ ] Remove console logs
- [ ] Add the list to the community page

**Comment by grmbyrn** [comment]
> Hi @grmbyrn, when you run it locally you get this error:
> 
> image
> 
> ```
>   ➜  Local:   http://localhost:5173/
>   ➜  Network: http://192.168.1.37:5173/
>   ➜  Network: http://192.168.1.33:5173/
>   ➜  press h + enter to show help
> Fetching data from API
> Rate limit: 60, Remaining: 0 10:46 AM
> Rate limit resets at: Mon Dec 02 2024 10:53:01 GMT+0000 (Western European Standard Time)
> Failed to fetch from https://api.github.com/orgs/torrust/repos: rate limit exceeded
> Error: Failed to fetch from https://api.github.com/orgs/torrust/repos
>     at Module.fetchWithCache (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/src/lib/utils/cache.ts:40:9)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async load (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/src/routes/(home)/+page.server.ts:13:16)
>     at async Module.load_server_data (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:61:17)
>     at async eval (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/page/index.js:140:13)
> Fetching data from API
> Rate limit: 60, Remaining: 0 10:46 AM
> Rate limit resets at: Mon Dec 02 2024 10:53:01 GMT+0000 (Western European Standard Time)
> Failed to fetch from https://api.github.com/orgs/torrust/repos: rate limit exceeded
> Error: Failed to fetch from https://api.github.com/orgs/torrust/repos
>     at Module.fetchWithCache (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/src/lib/utils/cache.ts:40:9)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async load (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/src/routes/(home)/+page.server.ts:13:16)
>     at async Module.load_server_data (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:61:17)
>     at async Promise.all (index 2)
>     at async Module.render_data (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/data/index.js:99:17)
>     at async resolve (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/respond.js:436:17)
>     at async Module.respond (/home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/runtime/server/respond.js:322:20)
>     at async file:///home/josecelano/Documents/git/committer/me/github/torrust/torrust-website/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:524:22
> ```
> 
> I guess it's because I restarted the server and the data is not in the store anymore. I think we should not show the contributors list in that case.
> 
> I would also add the list to the community page.
> 
> Finally, I would remove console logs for production, especially when it's only for debugging.
> 
> ### TODO
> * [ ]  Fix problem when get too many requests error but the store is empty. I think we could have a default value in this case with the list of contributor. We can update that default value from time to time.
> * [ ]  Remove console logs
> * [ ]  Add the list to the community page

> ### TODO
> * [X]  Fix problem when get too many requests error but the store is empty. I think we could have a default value in this case with the list of contributor. We can update that default value from time to time.
> * [X]  Remove console logs
> * [X]  Add the list to the community page

Hi @josecelano for the request error, I made an array of objects for the contributors and put it in `constants`. It will display the data coming from `constants` instead of the data coming from the GitHub API when the try catch block in `+page.server.ts` detects an error.

File: src/lib/utils/cache.ts [hunk 3]
```diff
@@ -4,6 +4,7 @@ import type { CacheEntry } from '$lib/utils/types';
 export const contributorsStore = writable<Map<string, CacheEntry<unknown>>>(new Map());
 
 const CACHE_TTL = 60 * 60 * 1000;
+const token = import.meta.env.VITE_GITHUB_TOKEN;
```
Comment (josecelano) [inline src/lib/utils/cache.ts]: Hi @grmbyrn why did you add the token again? I mean, it's fine for development, but if you add it, it should be optional for production, and the authorization header should not be included when it is empty.

File: src/lib/constants/constants.ts [hunk 4]
```diff
@@ -706,3 +706,263 @@ export const aboutTableData = [
 		ipfs: 'Cryptographic validation of all content'
 	}
 ];
+
+export const contributorsList = [
```
Comment (josecelano) [inline src/lib/constants/constants.ts]: Hi @grmbyrn I would rename `contributorsList` to `defaultContributorsList` or `staticContributorsList` or `initalContributorsList` to hint it would be fetched dinamically later.

**Review by josecelano — CHANGES_REQUESTED** [review]

File: src/lib/utils/cache.ts (hunk 3 above)
Comment (grmbyrn) [reply src/lib/utils/cache.ts]: Hi @josecelano I started deleting the console logs, then undid them to check something. I think I undid deleting the token without realising before I permanently deleted the console logs.

**Comment by josecelano** [comment]
ACK 8df5b468e90455f3806de58f5b9c7a82813d5a07
